docs/_blog/2025-10-02-recursive-hw.md

```diff
@@ -0,0 +1,336 @@
+current subtree to ensure it is within that range. If not we simply
+return the midpoint value (line 17) unchanged. This is one situation
+in which hardware recursion differs from software that we need to
+check this condition which adds range comparison at every node.
+
+Otherwise, in the 'hit' case (`invalidate`=`false`), we return '1'
+if the `way` is in the left subtree and '0' if it is in the right
+(line 18).
+
+Finally, we return a concatenation of the computed PLRU state vector
+(line 19).  Again note that the `invalidate=true` case simply reverses
+them meaning of '1' and '0'.
+
+We can see that maintaining a PLRU tree to support pseudo-LRU
+computation is quite similar in both software and hardware recursive
+forms. ROHD makes it simple to follow software patterns to generate
+hardware naturally when the computation is easily described
+recursively.
+
+## A Reduction Tree Component
+
+A reduction is an efficient arrangement of computation of associative
+operations (like sum or max) to compute a single result from an array
+of inputs.
+
+Reductions are common in both software and hardware and improve the
+latency of computation logarithmically by arranging a tree of
+computation to perform parallel operations even when the output is a
+single result. 
+
+In Hardware Description Languages (HDL)s, there is even operator
+syntax to perform common bit-level reductions like `or-reduction` and
+`and-reduction`. It is more difficult to describe reductions on more
+complex inputs or operations in traditional HDLs.
+
+Yet in software, more complex reduction trees are possible because of
+the compositional nature of software languages. For example the C++
+Standard Template Library (STL) contains a template operator
+`std::reduce` to perform reduction generically on STL
+containers. Indeed, these operations are threaded to ensure fastest
+possible execution in reduction tree arrangements within STL.
+
+To perform more complex reductions in hardware, the [ROHD Hardware
+Component Library (HCL)](<https://github.com/intel/rohd-hcl>) provides
+a `ReductionTree` component which takes an associative operation and
+populates a hardware tree arrangement to compute a hardware reduction
+with a latency that is logarithmic with the number of inputs. In
+hardware we need to consider pipelining as well for any significant
+reduction length.
+
+### Add Reduction Tree
+
+Here is an example of a reduction tree using the native add operations
+of SystemVerilog, but written using a ROHD generator class. 
+
+On lines 1-3 is the operation to be instantiated by tree generator, in
+this case a native addition of two inputs.
+
+On lines 9-10 is the tree generation, producing a binary tree of
+these 79 13-bit inputs and adding pipelining at every other level of
+adders.
+
+```dart
+ 1  Logic addReduce(List<Logic> inputs,
+ 2          {int depth, Logic? control, String name = ''}) =>
+ 3       inputs.reduce((v, e) => v + e);
```

Comment:
might I suggest not using the `reduce` function here as well, since it could conflate what you're building with what's built-in to Dart.  Maybe just `inputs[0] + inputs[1]`?  Or whatever conditionals are required to support 1 vs 2 inputs?  Even though it's more verbose, I think it makes the point more clearly, otherwise, one might ask "why not just use the `reduce` function you used here directly?"